cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -397,6 +446,97 @@ class FinalParameterUse extends UseImpl, TFinalParameterUse {
   }
 }
 
+class GlobalUse extends UseImpl, TGlobalUse {
+  Cpp::GlobalOrNamespaceVariable global;
+  IRFunction f;
+
+  GlobalUse() { this = TGlobalUse(global, f, ind) }
+
+  override FinalGlobalValue getNode() { result.getGlobalUse() = this }
+
+  override int getIndirection() { result = ind + 1 }
+
+  /** Gets the global variable associated with this use. */
+  Cpp::GlobalOrNamespaceVariable getVariable() { result = global }
+
+  /** Gets the `IRFunction` whose body is exited from after this use. */
+  IRFunction getIRFunction() { result = f }
+
+  final override predicate hasIndexInBlock(IRBlock block, int index) {
+    exists(ExitFunctionInstruction exit |
+      exit = f.getExitFunctionInstruction() and
+      block.getInstruction(index) = exit
+    )
+  }
+
+  override SourceVariable getSourceVariable() { sourceVariableIsGlobal(result, global, f, ind) }
+
+  final override Cpp::Location getLocation() { result = f.getLocation() }
+
+  /**
+   * Gets the type of this use after specifiers have been deeply stripped
+   * and typedefs have been resolved.
+   */
+  Type getUnspecifiedType() { result = global.getUnspecifiedType() }
+
+  override predicate isCertain() { any() }
+
+  override BaseSourceVariableInstruction getBase() { none() }
+}
+
+class GlobalDef extends TGlobalDef {
+  Cpp::GlobalOrNamespaceVariable global;
+  IRFunction f;
+  int indirectionIndex;
+
+  GlobalDef() { this = TGlobalDef(global, f, indirectionIndex) }
+
+  /** Gets the global variable associated with this definition. */
+  Cpp::GlobalOrNamespaceVariable getVariable() { result = global }
+
+  /** Gets the `IRFunction` whose body is evaluated after this definition. */
+  IRFunction getIRFunction() { result = f }
+
+  /** Gets the global variable associated with this definition. */
+  int getIndirectionIndex() { result = indirectionIndex }
+
+  /** Holds if this definition or use has index `index` in block `block`. */
+  final predicate hasIndexInBlock(IRBlock block, int index) {
```

Comment:
Should this also have an `override`? And can the QLDoc be dropped (as it is an override)?

Reply:
It's actually not an override because this class doesn't extend `DefImpl`. If `GlobalDef` _did_ extend `DefImpl` then your proposal is very much correct, but because a `DefImpl` currently _has_ to have an `Operand` representing the address of the definition a `GlobalDef` cannot be a `DefImpl` (since a `GlobalDef` is a kind of synthesized definition whose address isn't present in the IR).

The `Use` side of this has been refactored in a previous PR to handle this situation: We have a `UseImpl` class, and we have a `OperandBasedUse` to handle the common case that a `Use` has an operand that represents the address directly in the IR. But (unlike a `DefImpl`) a `UseImpl` doesn't have to be `OperandBasedUse`, and that's why the `GlobalUse` can extend `UseImpl` without requiring an `Operand`.

We could do a similar refactoring for `DefImpl` (i.e., 
remove the requirement that `DefImpl` has to have an `Operand`, and create a `OperandBasedDef` class and let most of the SSA definitions extend that class. Then we could have `GlobalDef` just extend the `DefImpl` class). It shouldn't be a difficult refactoring to do. I just chose to not do it in this PR to keep the diff as small as possible. But I'd be happy to do that refactoring if it helps you review the PR.

Reply:
Oh, right, I was mistakenly under the impression that it was extending more than `TGlobalDef`.

Reply:
No need to do the refactoring here.